src/ansys/dpf/composites/sampling_point_solid_stack.py

```diff
@@ -0,0 +1,794 @@
+# SOFTWARE.
+
+"""Implements sampling point functionality for a stack of solid elements."""
+from collections.abc import Collection, Sequence
+from typing import Any
+
+from ansys.dpf.core import UnitSystem
+from matplotlib.patches import Rectangle
+import numpy as np
+import numpy.typing as npt
+
+from ansys.dpf import core as dpf
+from ansys.dpf.composites.constants import (
+    FailureOutput,
+    Spot,
+    Sym3x3TensorComponent,
+    strain_component_name,
+    stress_component_name,
+)
+from ansys.dpf.composites.layup_info import (
+    AnalysisPlyInfoProvider,
+    ElementInfoProvider,
+    SolidStack,
+    SolidStackProvider,
+)
+from ansys.dpf.composites.solid_stack_results import (
+    get_through_the_thickness_failure_results,
+    get_through_the_thickness_results,
+)
+
+from ._sampling_point_helpers import (
+    add_ply_sequence_to_plot_to_sp,
+    add_results_to_plot_to_sp,
+    get_analysis_plies_from_sp,
+    get_data_from_sp_results,
+    get_indices_from_sp,
+    get_offsets_by_spots_from_sp,
+    get_ply_wise_critical_failures_from_sp,
+    get_polar_plot_from_sp,
+    get_result_plots_from_sp,
+)
+from .failure_criteria import CombinedFailureCriterion
+from .layup_info._layup_info import (
+    _get_layup_model_context,
+    get_material_names_to_dpf_material_index,
+)
+from .layup_info.material_operators import MaterialOperators
+from .layup_info.material_properties import get_material_metadata
+from .result_definition import FailureMeasureEnum
+from .sampling_point_types import FailureResult, SamplingPoint, SamplingPointFigure
+from .server_helpers import version_older_than
+from .unit_system import get_unit_system
+
+
+class SamplingPointSolidStack(SamplingPoint):
+    """Implements the ``Sampling Point`` object for a stack of solid elements.
+
+    This class provides for plotting the lay-up and results at a certain point of the
+    layered structure. The results, including ``analysis_plies``, ``e1``, ``s12``, and
+    ``failure_modes``, are always from the bottom to the top of the laminate (along
+    the element normal direction). Postprocessing results such as ``e1`` are returned
+    as flat arrays where ``self.spots_per_ply`` can be used to compute the index for
+    a certain ply.
+
+    Parameters
+    ----------
+    name :
+        Name of the object.
+    element_id :
+       Element label of a solid element. The sampling will automatically select to
+       full stack of solid elements where this element is located.
+
+    Notes
+    -----
+    The results of layered elements are stored per integration point. A layered shell element
+    has a number of in-plane integration points (depending on the integration scheme) and
+    typically three integration points through the thickness. The through-the-thickness
+    integration points are called `spots`. They are typically at the ``BOTTOM``
+    and ``TOP`` of the layer. This notation is used here to identify the corresponding data.
```

Comment:
On first reading this, I was confused why there are 3 integration points through the thickness, but then only `BOTTOM` and `TOP` are mentioned.

Maybe we should rephrase slightly, for example:

> They are typically at the ``BOTTOM``, ``MIDDLE``, and ``TOP`` of the layer. The ``MIDDLE`` spot is not available for solid elements, so only the ``BOTTOM`` and ``TOP`` spots are used to identify the corresponding data.

(not sure if I captured the intent behind "This notation is used here  to identify the corresponding data." fully)